src/threads/coro-1.1.0.pre2/coro.man
.na
.TH CORO 2 "1.1.0.pre2" "GNU" "C Coroutines"
.SH NAME
co_create, co_call, co_resume, co_delete, co_exit_to, co_exit \- C coroutine management

.SH SYNOPSIS
.nf
.B #include <coro.h>
.sp
.BI "struct coroutine {"
.nl
.BI "    ..."
.nl
.BI "    struct coroutine *caller;"
.nl
.BI "    struct coroutine *resumeto;"
.nl
.BI "    void *user;"
.nl
.BI "    ..."
.nl
.BI "};"
.sp
.BI "extern struct coroutine *co_current;"
.nl
.BI "extern struct coroutine co_main[];"
.sp
.BI "struct coroutine *co_create(void *" func ", void *" stack ", int " stacksize ");"
.nl
.BI "void co_delete(struct coroutine *" co ");"
.nl
.BI "void *co_call(struct coroutine *" co ", void *" data ");"
.nl
.BI "void *co_resume(void *" data ");"
.nl
.BI "void *co_exit_to(struct coroutine *" co ", void *" data ");"
.nl
.BI "void *co_exit(void *" data ");"
.SH DESCRIPTION
The
.B coro
library implements the low level functionality for coroutines.
For a definition of the term
.B coroutine
see
.IR "The Art of Computer Programming" " by " "Donald E. Knuth" .
In short, you may think of coroutines as a very simple cooperative
multitasking environment where the switch from one task to another
is done explicitly by a function call.  And, coroutines are fast.
Switching from one coroutine to another takes only a couple of
assembler instructions more than a normal function call.

This document defines an API for the low level handling of coroutines
i.e. creating and deleting coroutines and switching between them.
Higher level functionality (scheduler, etc.) is not covered.

.SS Functions
The following functions are defined:
.TP
.B co_create
This function creates a new coroutine.
.I func
is the entry point of the coroutine.  It will be called with one
arg, a
.BR "void *" ,
which holds the data given to the first
.B co_call
call.  If
.I func
terminates, an implicit
.B co_resume
with the function's return value is executed and the next
.BR co_call / co_resume
reenters the function at the top.

.I stack
is the base of the stack this coroutine will use and
.I stacksize
its size in bytes.  You may pass a
.B NULL
pointer for
.I stack
in which case the memory will be allocated by
.B co_create
itself.  Both,
.IR stack " and " stacksize
are aligned to system requirements.
A
.I stacksize
of less then 128 bytes will be rejected.
You have to make sure, that the stack is large enough for your
coroutine and possible signal handlers (see below).  The stack
will not grow!  (Exception: the
.B co_main
coroutine uses the standard system stack which may still grow.)

On success, a handle
.RB ( "struct coroutine *" )
for a new coroutine is returned, otherwise
.BR NULL .

.TP
.B co_delete
This function deletes the given coroutine
.IR co .
If the stack for this coroutine was allocated by
.B co_create
it will be freed.  After a coroutine handle was passed to
.B co_delete
it is invalid and may not be used any more.

It is invalid for a coroutine to delete itself with this
function.

.TP
.B co_call
This function passes execution to the given coroutine
.IR co .
The
.I data
pointer is transferred to the other coroutine.
The current coroutine is suspended until another one restarts it with a
.B co_call
or
.B co_resume
call.  At that moment the
.B co_call
returns with the data pointer received from the other coroutine.

Calling oneself returns immediately with the given
.IR data .

.TP
.B co_resume
This function passes execution back to the coroutine which either
initially started this one or restarted it after a prior
.BR co_resume .
The
.I data
pointer and the return value is the same as for
.BR co_call .

.TP
.B co_exit_to
This function does the same a
.B co_delete(co_current)
followed by a
.B co_call
would do.  That is, it deletes itself and then passes execution
and the
.I data
pointer to another coroutine
.IR co .

.TP
.B co_exit
This function does the same a
.B co_delete(co_current)
followed by a
.B co_resume
would do.  That is, it deletes itself and then passes execution
and the
.I data
pointer back to the coroutine which either initially started
this one or restarted it after a prior
.BR co_resume .

.SS Global variables
Two global read only variables are defined:
.TP
.B co_current
This variable always holds the handle of the currently active
coroutine.  It may be used to access the public fields of the
.BR "struct coroutine" .

.TP
.B co_main
This identifier may be used as a handle for the initial,
automagically created coroutine that performed the first
.B co_call
call.  Don't try to
.B co_delete
it *g*.

.SS Public struct coroutine fields
The following elements of the
.B struct coroutine
may be used by applications:
.TP
.B struct coroutine *caller;
This is the handle of the previously active coroutine.  Only valid
while the coroutine is active.
.TP
.B struct coroutine *resumeto;
This is the handle of the coroutine where
.BR co_resume " and " co_exit
would switch to.  Only valid while the coroutine is active.
.TP
.B void *user;
This field is reserved for applications.  It is set to
.BR NULL " by " co_create
and isn't touched until a
.B co_delete
invalidates the structure.

.SS Notes
Some interactions with other parts of the system are covered here.
.TP
.B Signals
First, a signal handler is not defined to run in any specific
coroutine.  The contents of the
.BR co_current " and " co_main
variables is undefined and you may not call
.B any
coroutine function.  The only way to leave the signal handler is
by a return statement.

Second, the signal handler may run with the stack of any coroutine,
even with the stack of library internal coroutines which have an
undefined stack size (just enough to perform a kernel call).
Using and alternate stack for signal processing (see
.BR sigaltstack (2))
is recommended!

Conclusion: avoid signals like a plague.  The only thing you may
do reliable is setting some global variables and return.
Simple kernel calls may work too, but nowadays it's pretty hairy
to tell, which function really is a kernel call.
(Btw, all this applies to normal C programs, too.  The coroutines
just add one more problem *g*)
.TP
.B setjmp/longjmp
The use of
.BR setjmp / longjmp
is limited to jumping inside one coroutine.  Never try to jump from
one coroutine to another with
.BR longjmp .
.TP
.B Passing local data
With coroutines there is a new way to generate dangling references.
If one coroutine passes a pointer to local data (storage class auto)
to another coroutine this pointer
.B may
become invalid at the moment that coroutine is restarted or deleted.
Especially, passing a pointer to auto data to
.BR co_exit / co_exit_to
is only valid, if the stack wasn't allocated by
.BR co_create .
.TP
.B Minimalism
The only necessary function would be
.BR co_create ", " co_delete ", and " co_call .
The other functions
.RB ( co_resume ", " co_exit ", and " co_exit_to )
could easily be implemented in a portable way.  But they seemed
fundamental enough that I added them to the core library.

.SH DIAGNOSTICS
Some fatal errors are caught by the library.  If one occurs,
a short message is written to file descriptor 2 (stderr) and
a segmentation violation is generated.
.TP
.B coro: coroutine deletes itself
A coroutine has called
.B co_delete
with it's own handle.
.TP
.B coro: resume to deleted coroutine
A coroutine has deleted itself with
.BR co_exit " or " co_exit_to
and the coroutine that was activated by the exit tried a
.BR co_resume .
.TP
.B coro: stale coroutine called
Someone tried to active a coroutine that has already been
deleted.  This error is only detected, if the stack of the
deleted coroutine is still resident in memory.

.SH SEE ALSO
A good book about coroutines *g*

.SH BUGS
There are no known bugs.  But, this library is still
in development.  Especially, it's only implemented
for x86/Linux and FreeBSD yet.  Later implementations
for other architectures may require some redefinitions.

Obbug: These routines should be in libc *g*

Bug reports and comments to
.BR froese@gmx.de .


src/CHANGES

----------------------------------------------------------------------
This file is intended to help track internal changes to the library.
Latest changes should be added to the top of the file.
----------------------------------------------------------------------

2004/03/29 - zf
  * read_proc_file() is using close() to close the proc file
    Because it is called from the scheduler, it should use the
    SYS_close syscall directly.

2003/08/11 - zf
  * Fastpath improvement for buffer-cache hitting disk requests
  
2003/08/09 - zf
  * New eepoll patch for 2.6.0-test2-bk8

2003/08/07 - zf
  * Reformatted micro-benchmark results
  * Save errno in thread stack to avoid conflict with NPTL
  * Reduced I/O polling rate
  * Added group submit for nio. Default batch factor is 32.
    Tunable by IOQ_DEPTH_MAX in nio.c

2003/08/01 - zf
  * Added pipe scalability benchmark at src/bench/pipetest_thread.

2003/07/30 - jrvb
  * Added util/config.[ch], to clean up runtime options
  * Made several compile-time options to runtime options 

2003/07/28 - zf
  * Added new concurrent I/O benchmark p_bench_concurrent_io.c

2003/07/25 - jrvb
  * Changed init code in a number of places to use gcc's
    __attribute__((constructor)).  This is much more reliable than
    trying to do things by hand.  NOTE: in a few cases, things are
    still done by hand, eg. where order matters.
  * added diskio_immediate.c, to do immediate, blocking disk IO.  

2003/07/24 - zf
  * Pipe benchmark in bench/p_bench_pipe.c

2003/07/23 - jrvb
  * added some crude support for dynamic linking of libpthread.so to aio/Makefile  

2003/07/22 - jrvb
  * tweak Makefile in bench to kill orphaned capriccio kernel threads
  * fixed program name in bench/test_readdir.c

2003/07/21 - zf
  * automate benchmarking process of bench/ 
  * cosmetic changes to SYS_SET_SYSCALL (gcc 3.3+ complaining about 
    assigning const var to non-const)
  * nptl benchmarking
  * nptl build notes in doc/

2003/07/19 - zf
  * numerous signal fixes
  * signal test cases in bench/
  * apache now runs with no patching at all!
    ("-D DEBUG" command line argument still needed)

2003/07/15 - zf
  * (untested) full signal implementation in signal.c
  * fixed bug that stops pthread_exit() from exiting the main thread
  * now we pass most of test cases in bench/.  several we cannot do
    are either because we are cooperative or we do not have support some
    optional functionalities.

2003/07/11 - zf
  * __errno_location() now returns a thread-local error number instead
    of a global one
    
  * (untested) new select() wrapper in nio.c (not in default build)
  
  * Apache now works with nio (with Rob's configure command line and
    a slightly modified apache patch(in cvs))

2003/07/09 - zf
  * Extended epoll kernel patch in kernel/patches/eepoll-2.5.70.diff
    It's applicable to kernel 2.5.70 (or up hopefully)

2003/07/08 - zf
  * checked in i/o wrapping code using the extended epoll kernel interface
    Code is in nio/ and not used by default.  Add 'USE_IO_LIB=nio' to
    Make.opts to use it instead of the aio/ code.
    ( This is still preliminary and overlaps with the aio/ code.  
      It should be merged with aio/ code later.)

  * added CAP_NODEBUG env. var to disable debug messages at runtime.

  * added tests and benchmarks in bench, mainly taken from MIT pthread

2003/06/18 - jrvb
  * checked in initial (not tested) support for send/recv and friends
    to blocking_io.c

2003/04/24 - jrvb
  * switched util/occ_list.c to use spinlocks, instead of true
    optomistic concurrency control.
  * fixed kthread disk IO
  * made diskio_kthread the default
  * added more reporting to knot

2003/04/19 - jrvb
  * added line counts to makefiles

2003/03/24 - jrvb
  * have sockio_poll check IO first - faster w/ knot 

2003/03/23 - zf
  * fixes to assert()'s with side-effects.  the side-effects will not be
    executed when asserts are disable, causing potential bugs

2003/03/23 - jrvb
  * bugfixes all over the place
  * change timers, counters, etc to show diff b/w process time and
    real time when perfctr is available
  * insert yields in various places before IO calls, for better
    blocking graph and to avoid long edges
  * add tracking of which syscall we are in
  * ugly hacks to avoid init races w/ perfctr and IO libs
  * signal changes for better core dumps
  * bugfixes to diskio_kthread
  * many fixes to edge length tracking & stats gathering
  * fixes & improvements to scheduling, including basic admission control

2003/03/21 - jrvb
  * cleanup of blocking graph code
    - prepare for final integration w/ cil blocking graph
    - reduce overhead through sampling
  * fixes to cpu timing code & resource counting

2003/03/21 - zf
  * perfctr-based timing in timing.c, set USE_PERFCTR = 1 in Make.opts to
    use.  Then kill -sigusr1 <pid> will print out timing info

2003/03/19 - jrvb
  * optomistic concurrency control list structures are working w/
    diskio_kthread and linux clone()
  * make IO regime runtime selectable via environment vars:
       CAPRICCIO_SOCKIO and CAPRICCIO_DISKIO
  * change current_usec() to current_usecs() - for some reason the
    java programmer in me always wants the 's' from currentTimeMillis()
  * modify polling behavior in scheduler
    - if there are no active threads, allow sockio to block
    - max block time == min sleep time for sleeping threads
  * change src/aio/Makefile behavior - looks now for make variables
    HAVE_AIO and HAVE_SYS_EPOLL.  These should be defined in src/Make.opts
  * make debug selectable from w/in Make.opts.  See Make.opts.sample
    for info.

2003/03/18 - zf
  * timing routines using rdtsc in util/timing.h
    these can be used to measure our overhead of polling/scheduling

2003/03/15 - zf
  * updated apache patch - more signal ignoring
  * fix poll() race bug - we may return wrong result and corrupt the poll list
  * use backtrace_symbols() anyway because it gives us some function names

src/threads/coro-1.1.0.pre2/sample/coro.c
/*
    Sample code for coroutine implementation.

    The main porting points:

    - How to allocate the stack?  mmap or malloc or something else.
      Requires adjustments in co_create and co_delete.
    - Check fatal.  It should generate a segv or something that
      is easily caught in a debugger.
    - Implement co_call.
    - Implement __co_wrap.
    - Adjust co_create to setup the stack for the first co_call to call wrap.

    Maybe you have to implement parts (co_call, __co_wrap) in an .s file.
*/

#include <unistd.h>		/* for write, mmap, munmap */
#include <sys/mman.h>		/* for mmap, munmap */
#include "coro.h"		/* for struct coroutine */

/* some compiler checks */
#if !defined(i386)
#error For x86-CPUs only
#endif
#if !defined(__GNUC__)
#warning May break without gcc.  Be careful.
#endif

#ifndef PAGESIZE
#define PAGESIZE	4096
#endif

struct coroutine co_main[1] = { { 0 } };
struct coroutine *co_current = co_main;


/* make sure that fatal generates a segv.  it mustn't return! */
#define fatal(msg) \
    for (;;)						\
    {							\
	write(2, "coro: " msg "\r\n", sizeof(msg)+7);	\
	*(unsigned int *)0 = 0xfee1dead;		\
    }



/*
    Create new coroutine.
    'func' is the entry point
    'stack' is the start of the coroutines stack.  if 0, one is allocated.
    'size' is the size of the stack
*/


/* !!! special calling convention. !!! */
static void
__co_wrap(void *data) /* arg in the register where results are returned */
{
    co_current->resumeto = co_current->caller;

    for (;;)
	data = co_resume(co_current->func(data));
}

/* !!! use mmap or malloc.  setup stack for co_call. !!! */
struct coroutine *
co_create(void *func, void *stack, int size)
{
    struct coroutine *co;
    int to_free = 0;

    if (size < 128)
	return 0;

    if (stack == 0)
    {
	/* align stacksize */
	size += PAGESIZE-1;
	size &= ~(PAGESIZE-1);
#if USE_MMAP
	stack = mmap(0, size, PROT_READ|PROT_WRITE,
			      MAP_PRIVATE|MAP_ANON, -1, 0);
	if (stack == (void*)-1)
	    return 0;
#else /* malloc */
	size -= 32;			/* some place for malloc to waste */
	stack = malloc(size);
	if (stack == 0)
	    return 0;
#endif
	to_free = size;
    }
    co = stack + size;
    (unsigned long)co &= ~3;	// align
    co -= 1;

    /* initialize struct coroutine */
    co->sp = co;
    co->caller = 0;
    co->resumeto = 0;
    co->user = 0;
    co->func = func;
    co->to_free = to_free;

    /*
	Setup stack so that a pop_register_vars and a return starts
	the wrapper.  It is started by a return, so its single arg
	is passed in the return register.
    */
    *--(void **)co->sp = __co_wrap;	// return addr (here: start addr)
    *--(void **)co->sp = 0;		// reg1
    *--(void **)co->sp = 0;		// reg2
    *--(void **)co->sp = 0;		// ...
    return co;
}



/*
    delete a coroutine.
*/

void
co_delete(struct coroutine *co)
{
    if (co == co_current)
	fatal("coroutine deletes itself");

    if (co->to_free)
#if USE_MMAP
	munmap((void*)co + sizeof(*co) - co->to_free, co->to_free);
#else /* malloc */
	free((void*)co + sizeof(*co) - co->to_free);
#endif
}



/*
    delete self and switch to 'new_co' passing 'data'
*/

static void *helper_args[2];

static void
del_helper(void **args)
{
    for (;;)
    {
	if (args != helper_args)
	    fatal("resume to deleted coroutine");
	co_delete(co_current->caller);
	args = co_call(args[0], args[1]);
    }
}

void
co_exit_to(struct coroutine *new_co, void *data)
{
    static struct coroutine *helper = 0;
    static char stk[256];

    helper_args[0] = new_co;
    helper_args[1] = data;

    if (helper == 0)
	helper = co_create(del_helper, stk, sizeof(stk));

    /* we must leave this coroutine.  so call the helper. */
    co_call(helper, helper_args);
    fatal("stale coroutine called");
}

void
co_exit(void *data)
{
    co_exit_to(co_current->resumeto, data);
}



/*
    Call other coroutine.
    'new_co' is the coroutine to switch to
    'data' is passed to the new coroutine
*/

/* !!! pseudo code !!! */
void *
co_call(struct coroutine *new_co, void *data)
{
    push_all_register();
    co_current->sp = stack_pointer;

    new_co->caller = co_current;
    co_current = new_co;

    stack_pointer = co_current->sp;
    pop_all_registers();
    return data;
}

void *
co_resume(void *data)
{
    data = co_call(co_current->resumeto, data);
    co_current->resumeto = co_current->caller;
    return data;
}
